File: php/0116_populating_next_right_pointers_in_each_node.php
<?php
// Runtime: 35 ms
// Memory Usage: 19.4 MB

/**
 * 
 * 
 * Definition for a Node.
 * class Node {
 *     function __construct($val = 0) {
 *         $this->val = $val;
 *         $this->left = null;
 *         $this->right = null;
 *         $this->next = null;
 *     }
 * }
 */

class Solution {
    /**
     * @param Node $root
     * @return Node
     */
    public function connect($root) {
        if ($root === null || $root->left === null) {
            return $root;
        }
        
        $root->left->next = $root->right;
        if ($root->next !== null) {
            $root->right->next = $root->next->left;
        }
        
        $root->left = $this->connect($root->left);
        $root->right = $this->connect($root->right);
        
        return $root;
    }
}

File: go/0116_populating_next_right_pointers_in_each_node.go
// Runtime: 4 ms
// Memory Usage: 6.5 MB

/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Left *Node
 *     Right *Node
 *     Next *Node
 * }
 */

func connect(root *Node) *Node {
	if root == nil || root.Left == nil {
		return root
	}

	root.Left.Next = root.Right
	if root.Next != nil {
		root.Right.Next = root.Next.Left
	}

	root.Left = connect(root.Left)
	root.Right = connect(root.Right)

	return root
}